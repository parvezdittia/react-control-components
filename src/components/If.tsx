import React, { FC, ReactNode } from 'react';

export interface IfProps {
  condition: boolean;
  children: ReactNode;
}

const If: FC<IfProps> = ({ condition, children }) => {
  if (condition) {
    return <>{children}</>;
  }
  return null;
};

export default If;
