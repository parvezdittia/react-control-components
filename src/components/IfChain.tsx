import {
  FC,
  Children,
  isValidElement,
  cloneElement,
  ReactElement,
} from 'react';
import If, { IfProps } from './If';

interface IfChainProps {
  children: Array<ReactElement<IfProps>> | ReactElement<IfProps>;
}

const IfChain: FC<IfChainProps> = ({ children }) => {
  let childToReturn: ReactElement<IfProps> | null = null;

  Children.forEach(children, (child) => {
    if (childToReturn) return;
    if (!isValidElement(child)) {
      throw new Error(`${child} is not a valid React element`);
    } else if (child.type !== If) {
      throw new Error(`${child} is not an <If /> component`);
    } else {
      let result = child.props.condition;
      if (result) {
        childToReturn = cloneElement(child, {
          condition: result,
        });
      }
    }
  });

  return childToReturn;
};

export default IfChain;
