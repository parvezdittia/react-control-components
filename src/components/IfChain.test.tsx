import React from 'react';
import IfChain from './IfChain';
import If from './If';
import renderer from 'react-test-renderer';

describe('IfChain component', () => {
  it('renders without errors when all children are If components', () => {
    const tree = renderer
      .create(
        <IfChain>
          <If condition={true}>Child 1</If>
          <If condition={false}>Child 2</If>
          <If condition={true}>Child 3</If>
        </IfChain>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('throws an error when an invalid child component is present', () => {
    // Mock console.error to suppress error logging in the console
    //console.error = jest.fn();

    const tree = renderer
      .create(
        <IfChain>
          <If condition={true}>Child 1</If>
          <div>Invalid Child</div>
          <If condition={true}>Child 2</If>
        </IfChain>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();

    // Restore console.error
    //console.error = consoleError;
  });

  // it('renders the correct number of valid child components', () => {
  //     const { container } = render(
  //         <IfChain>
  //             <If condition={true}>Child 1</If>
  //             <If condition={false}>Child 2</If>
  //             <If condition={true}>Child 3</If>
  //             <If condition={true}>Child 4</If>
  //         </IfChain>
  //     );
  //     expect(container.children.length).toBe(3); // Only 3 valid If components
  // });
});
