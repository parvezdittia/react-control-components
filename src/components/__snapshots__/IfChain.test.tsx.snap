// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IfChain component renders without errors when all children are If components 1`] = `
<div>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
</div>
`;

exports[`IfChain component throws an error when an invalid child component is present 1`] = `
<div>
  <div>
    1
  </div>
  <div>
    2
  </div>
</div>
`;
